<a href="https://colab.research.google.com/github/Adeela-saleem/MachineLearning_Algorithms/blob/main/Supervised_Learning/Regression/MLR_Scratch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np


In [2]:
from sklearn.datasets  import load_diabetes

In [3]:
X,y=load_diabetes(return_X_y=True)

In [5]:
X
y

array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
        69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
        68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
        87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
       259.,  53., 190., 142.,  75., 142., 155., 225.,  59., 104., 182.,
       128.,  52.,  37., 170., 170.,  61., 144.,  52., 128.,  71., 163.,
       150.,  97., 160., 178.,  48., 270., 202., 111.,  85.,  42., 170.,
       200., 252., 113., 143.,  51.,  52., 210.,  65., 141.,  55., 134.,
        42., 111.,  98., 164.,  48.,  96.,  90., 162., 150., 279.,  92.,
        83., 128., 102., 302., 198.,  95.,  53., 134., 144., 232.,  81.,
       104.,  59., 246., 297., 258., 229., 275., 281., 179., 200., 200.,
       173., 180.,  84., 121., 161.,  99., 109., 115., 268., 274., 158.,
       107.,  83., 103., 272.,  85., 280., 336., 281., 118., 317., 235.,
        60., 174., 259., 178., 128.,  96., 126., 28

Here is a clear and concise summary in English that you can add as a text cell in your Jupyter / Google Colab notebook:

---

### **Why Load Datasets Directly as NumPy Arrays (`return_X_y=True`) Instead of a Pandas DataFrame?**

1. **Efficiency for Mathematical Operations:**
* When implementing ML algorithms from scratch (e.g., Multiple Linear Regression), operations like matrix multiplication ($X^T X$), transpositions, and inversions are executed much faster and more directly using **NumPy arrays**.


2. **Scikit-Learn Compatibility:**
* Scikit-Learn models (like `LinearRegression().fit(X, y)`) natively process NumPy arrays. Passing a DataFrame internally converts it to a NumPy array anyway.


3. **No Data Cleaning Needed:**
* Pre-packaged benchmark datasets (like `load_diabetes()`) are already clean, missing-value-free, scaled, and purely numerical. Therefore, Pandas data-wrangling features (`df.fillna()`, `df.describe()`) are not required.



> **When to use Pandas instead?** Use a Pandas DataFrame when performing Exploratory Data Analysis (EDA), inspecting column names, handling categorical data, or cleaning raw real-world datasets.

In [6]:
X.shape

(442, 10)

In [7]:
y.shape

(442,)

Using sklearn's Linear regression

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=2)

In [10]:
X_train.shape

(353, 10)

In [11]:
X_test.shape

(89, 10)

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
reg=LinearRegression()

In [14]:
reg.fit(X_train,y_train)

LinearRegression()

In [15]:
y_pred = reg.predict(X_test)

In [16]:
from sklearn.metrics import r2_score

In [17]:

r2_score(y_test,y_pred)

0.4399338661568968

In [19]:
reg.coef_

array([  -9.15865318, -205.45432163,  516.69374454,  340.61999905,
       -895.5520019 ,  561.22067904,  153.89310954,  126.73139688,
        861.12700152,   52.42112238])

Here is a clean English summary of these two concepts that you can paste directly as a Markdown cell into your notebook:

---

### **1. Understanding Model Accuracy ($R^2 \approx 0.44$ / $44\%$)**

* **Reason for Lower $R^2$ Score:** Unlike artificial/synthetic datasets (which yield high $R^2$ scores), the **Diabetes dataset** consists of real-world clinical data. Medical data contains significant natural variance, noise, and complex relationships.
* **Model Limitation:** Multiple Linear Regression assumes a strictly linear relationship between input features and target variables. Real-world medical variables often exhibit non-linear interactions that linear models cannot fully capture.
* **Benchmarking:** A score of **$\sim 44\%$** is standard for a simple Linear Regression model on this dataset. Advanced techniques like regularization (Ridge/Lasso) or non-linear algorithms (e.g., Decision Trees, Random Forest) are required to boost accuracy.

---

### **2. Interpretation of Regression Coefficients (`reg.coef_`)**

Linear Regression predicts values using the formula:

$$y = \text{Intercept} + \beta_1 X_1 + \beta_2 X_2 + \dots + \beta_{10} X_{10}$$

The `reg.coef_` array contains 10 values ($\beta_1 \dots \beta_{10}$), corresponding to the **10 features** in the dataset:

* **Positive Coefficients (e.g., $+861.12$, $+516.69$):** Represent a **direct relationship**. As these feature values increase, disease progression increases.
* **Negative Coefficients (e.g., $-895.55$, $-205.45$):** Represent an **inverse relationship**. As these feature values increase, disease progression decreases.
* **Magnitude (Absolute Value):** Indicates feature importance. Features with larger numerical values (e.g., $-895.55$ and $+861.12$) exert the strongest influence on the model's predictions, whereas smaller values (e.g., $-9.15$ or $+52.42$) have minimal impact.

In [18]:

reg.intercept_

np.float64(151.88331005254167)

# Making our own Linear Regression Class

Here is a clear and concise summary of the `ownClass` implementation in English that you can add as a Markdown text cell in your notebook:

---

### **Custom Multiple Linear Regression (`ownClass`) Implementation Summary**

1. **Constructor (`__init__`):**
* Initializes `self.coef_` (weights/slopes) and `self.intercept_` (bias/$Y$-intercept) as `None`.
* Follows Scikit-Learn convention by using trailing underscores for parameters learned during model training.


2. **Inserting the Bias/Dummy Column (`np.insert`):**
* **Formula:** Multiple Linear Regression equation is $y = \beta_0(1) + \beta_1 X_1 + \dots + \beta_n X_n$.
* **Purpose:** Adds a column of ones ($1\text{s}$) at index `0` of `X_train` (`axis=1`). This allows the intercept term ($\beta_0$) to be computed seamlessly alongside feature coefficients ($\beta_1 \dots \beta_n$) using matrix operations (Normal Equation: $\boldsymbol{\beta} = (\mathbf{X}^T \mathbf{X})^{-1} \mathbf{X}^T \mathbf{y}$).

In [22]:
class ownClass:

  def __init__(self):
      self.coef_=None
      self.intercept_=None
  def fit(self,X_train,y_train):
    X_train=np.insert(X_train,0,1,axis=1)

    #calculate coeffs
    betas=np.linalg.inv(np.dot(X_train.T,X_train)).dot(X_train.T).dot(y_train)
    self.intercept_=betas[0]
    self.coef_=betas[1:]

  def predict(self,X_test):
    y_pred=np.dot(X_test,self.coef_)+self.intercept_
    return y_pred




### **Mathematical Formulation of OLS (Normal Equation)**

1. **Normal Equation Computation:**
   * Uses the closed-form formula $\boldsymbol{\beta} = (\mathbf{X}^T \mathbf{X})^{-1} \mathbf{X}^T \mathbf{y}$ via `np.linalg.inv()` and `.dot()` to find optimal parameters in a single step without Gradient Descent.

2. **Extracting Intercept and Coefficients:**
   * **`self.intercept_ = betas[0]`**: The first element represents the bias/intercept ($\beta_0$) corresponding to the inserted dummy column of ones.
   * **`self.coef_ = betas[1:]`**: The remaining values represent feature weights ($\beta_1 \dots \beta_n$).

In [23]:
lr=ownClass()

In [24]:
lr.fit(X_train,y_train)

In [25]:
X_train.shape

(353, 10)

In [26]:
X_test.shape

(89, 10)

In [27]:
y_pred = lr.predict(X_test)

In [28]:
r2_score(y_test,y_pred)

0.43993386615689634

In [29]:
lr.coef_

array([  -9.15865318, -205.45432163,  516.69374454,  340.61999905,
       -895.5520019 ,  561.22067904,  153.89310954,  126.73139688,
        861.12700152,   52.42112238])

In [30]:

lr.intercept_

np.float64(151.88331005254165)

Here is a comprehensive, production-ready Markdown summary designed to be placed at the very end of your notebook as a final conclusion/wrap-up cell:

---

### **Notebook Conclusion & Comparison Summary**

#### **1. Objective**

The goal of this notebook was to implement **Multiple Linear Regression from scratch** using the OLS (Ordinary Least Squares) Normal Equation and validate its performance against Scikit-Learn’s built-in `LinearRegression` model using the **Diabetes Dataset**.

---

#### **2. Performance Comparison: Custom Class vs. Scikit-Learn**

| Metric / Parameter | Scikit-Learn `LinearRegression` | Custom Class (`ownClass`) | Match Status |
| --- | --- | --- | --- |
| **$R^2$ Score** | `0.4399338661568968` | `0.43993386615689634` | **Identical** ($\approx 0.44$) |
| **Intercept ($\beta_0$)** | `151.88331005254167` | `151.88331005254165` | **Identical** |
| **Coefficients ($\beta_1 \dots \beta_{10}$)** | Exact array match | Exact array match | **Identical** |

> **Key Takeaway:** Our scratch implementation achieves **floating-point precision parity** with Scikit-Learn's library implementation, proving that our matrix algebra ($\boldsymbol{\beta} = (\mathbf{X}^T \mathbf{X})^{-1} \mathbf{X}^T \mathbf{y}$) is 100% mathematically accurate.

---

#### **3. Core Technical Learnings**

1. **Dummy Feature Insertion (`np.insert`):**
* Prepending a column of $1\text{s}$ at index `0` of the feature matrix allows the bias/intercept ($\beta_0$) to be calculated simultaneously with feature weights ($\beta_1 \dots \beta_{n}$) in a single closed-form matrix operation.


2. **Dataset Selection & Model Benchmarking:**
* An $R^2$ score of $\sim 44\%$ is normal for simple Linear Regression on the Diabetes dataset due to complex, non-linear relationships present in real-world clinical data.


3. **Coefficient Interpretation:**
* Features with large positive coefficients (e.g., $+861.12$, $+516.69$) strongly drive up target progression, whereas negative coefficients (e.g., $-895.55$) exert an inverse relationship.